File: java/src/shortener/Shortener.java
package shortener;

import java.util.HashMap;
import java.util.concurrent.ThreadLocalRandom;

public class Shortener {
    private HashMap<String, ShortenedURL> urls;

    public Shortener() {
        urls = new HashMap<>();
    }

    public String shorten(String url) {
        ShortenedURL shortenedURL = new ShortenedURL(url);
        int seed = ThreadLocalRandom.current().nextInt();
        String hash = Base62.encode(seed);
        urls.put(hash, shortenedURL);

        return hash;
    }

    public ShortenedURL fetch(String hash) {
        if (!urls.containsKey(hash)) {
            return null;
        }

        ShortenedURL shortenedURL = urls.get(hash);
        shortenedURL.hit();
        return shortenedURL;
    }
}

File: kotlin/src/shortener/Shortener.kt
package shortener

import java.util.HashMap
import java.util.concurrent.ThreadLocalRandom

class Shortener {
    private val urls = HashMap<String, ShortenedURL>()

    fun shorten(url: String): String {
        val shortenedURL = ShortenedURL(url)
        val seed = ThreadLocalRandom.current().nextInt()
        val hash = Base62.encode(seed)
        urls.put(hash, shortenedURL)

        return hash
    }

    fun fetch(hash: String): ShortenedURL? {
        val shortenedURL = urls[hash]
        shortenedURL?.hit()
        return shortenedURL
    }
}
